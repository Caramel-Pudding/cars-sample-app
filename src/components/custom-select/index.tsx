import React, { FC, memo, useState, useRef, KeyboardEvent } from "react";
import classnames from "classnames";
import { useCloseOnOutsideClick } from "../../hooks/use-close-on-outside-click";
import { sharedClasses } from "../../consts/css";

import styles from "./styles.module.css";

export const inputDataTestId = "INPUT_DATA_TEST_ID";

export interface CustomSelectProps {
  readonly labelText: string;
  readonly options: string[];
  readonly chosenOption: string;
  readonly handler: (select: string) => void;
  readonly placeholder?: string;
  readonly containerClasses?: string;
  readonly selectClasses?: string;
  readonly optionClasses?: string;
}

export const CustomSelect: FC<CustomSelectProps> = memo(
  ({
    labelText,
    options = [],
    chosenOption,
    handler,
    placeholder = "",
    containerClasses = "",
    selectClasses = "",
    optionClasses = "",
  }) => {
    const [isOpen, setIsOpen] = useState(false);
    const list = useRef<HTMLUListElement>(null);
    useCloseOnOutsideClick(setIsOpen, isOpen, list);

    const toggleIsOpened = () => {
      setIsOpen((prevState) => !prevState);
    };

    const keyboardHandler = (event: KeyboardEvent) => {
      if (event.key === "Enter") {
        toggleIsOpened();
      }
    };

    const isElementSelected = (value: string) => chosenOption === value;

    const selectHandler = (value: string) => {
      if (value === chosenOption) {
        if (!placeholder) {
          return;
        }
        handler("");
        return;
      }
      handler(value);
    };

    return (
      <label className={classnames(styles.container, containerClasses)}>
        {labelText}
        <button
          data-testid={inputDataTestId}
          tabIndex={0}
          type="button"
          className={classnames(
            styles.input,
            sharedClasses.fonts.Md,
            selectClasses
          )}
          onClick={toggleIsOpened}
          onKeyDown={keyboardHandler}
        >
          {chosenOption || placeholder}
          <svg
            className={classnames(styles.bullet, {
              [styles.bulletActive]: isOpen,
            })}
            height="15"
            tabIndex={0}
            width="15"
            onClick={toggleIsOpened}
            onKeyDown={keyboardHandler}
          >
            <polygon points="1,1 7,10 14,1" style={{ fill: "#EDEDED" }} />
          </svg>
        </button>
        {isOpen && (
          <ul className={styles.list} ref={list}>
            {options.map((option) => (
              <li
                key={option}
                className={classnames(
                  styles.listElement,
                  isElementSelected(option)
                    ? styles.listElementSelected
                    : styles.listElementNotSelected,
                  optionClasses
                )}
                value={option}
              >
                <button
                  className={classnames(sharedClasses.fonts.Md, styles.button)}
                  type="button"
                  onClick={() => selectHandler(option)}
                >
                  {option}
                </button>
              </li>
            ))}
          </ul>
        )}
      </label>
    );
  }
);
