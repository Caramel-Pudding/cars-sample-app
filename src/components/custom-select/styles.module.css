.container {
    position: relative;
}

.input {
    border: solid 1px var(--light);
    padding: var(--offSm);
    margin-top: var(--offSm);
}

.list {
    position: absolute;
    top: 60px;
    background-color: white;
    z-index: 1;
    width: 100%;
}

.listElement {
    border: solid 1px var(--light);
    transition: background-color 0.2s ease-in;
}

.listElementSelected {
    background-color: rgba(234, 127, 40, 1);
    color: white;
}
  
.listElementNotSelected:hover {
    background-color: rgba(234, 127, 40, 0.75);
}
  
.listElementNotSelected:focus {
    background-color: rgba(234, 127, 40, 0.75);
}

.button {
    padding: var(--offSm);
    width: 100%
}