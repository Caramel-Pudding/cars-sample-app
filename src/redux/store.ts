import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { reducer as filtersReducer } from "./features/filters/slice";
import { reducer as paginationReducer } from "./features/pagination/slice";
import { reducer as sortingReducer } from "./features/sorting/slice";
import { reducer as carsReducer } from "./features/cars/slice";

export const store = configureStore({
  reducer: {
    filters: filtersReducer,
    pagination: paginationReducer,
    sorting: sortingReducer,
    cars: carsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = Readonly<ReturnType<typeof store.getState>>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
