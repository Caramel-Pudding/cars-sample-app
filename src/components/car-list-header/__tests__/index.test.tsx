import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";

import { CarListHeader } from "..";
import { store } from "../../../redux/store";
import { inputDataTestId } from "../../custom-select";

describe("CarListHeader", () => {
  const TestWrapper = (
    <Provider store={store}>
      <Router>
        <CarListHeader currentPageCarsCount={10} totalCarsCount={100} />
      </Router>
    </Provider>
  );

  afterEach(cleanup);

  it("Renders with proper data", () => {
    const { container, getByText } = render(TestWrapper);
    const carsInfo = getByText("Showing 10 of 100 results");
    expect(carsInfo).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
