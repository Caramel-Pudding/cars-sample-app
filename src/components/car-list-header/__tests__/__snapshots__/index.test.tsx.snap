// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarListHeader Renders with proper data 1`] = `
<div>
  <section
    class="header"
  >
    <section>
      <h1
        class="robotoLgBold headTitle"
      >
        Available Cars
      </h1>
      <div
        class="resultsSubTitle robotoMd"
      >
        Showing 
        10
         of 
        100
         results
      </div>
    </section>
    <section>
      <label
        class="container selectContainer"
      >
        Sort by
        <button
          class="input robotoMd selectInput"
          data-testid="INPUT_DATA_TEST_ID"
          tabindex="0"
          type="button"
        >
          Ascending
          <svg
            class="bullet"
            height="15"
            tabindex="0"
            width="15"
          >
            <polygon
              points="1,1 7,10 14,1"
              style="fill: #EDEDED;"
            />
          </svg>
        </button>
      </label>
    </section>
  </section>
</div>
`;
